packages/bull-vault/src/AuctionBull.sol

```diff
@@ -0,0 +1,565 @@
+        uint24 _poolFee
+    ) external {
+        require(msg.sender == auctionManager, "AB0");
+
+        if (_isSellingUsdc) {
+            // swap USDC to WETH
+            _exactInFlashSwap(
+                usdc,
+                weth,
+                _poolFee,
+                _usdcAmount,
+                _wethLimitAmount,
+                uint8(FLASH_SOURCE.LEVERAGE_REBALANCE_INCREASE_DEBT),
+                ""
+            );
+        } else {
+            // swap WETH to USDC
+            _exactOutFlashSwap(
+                weth,
+                usdc,
+                _poolFee,
+                _usdcAmount,
+                _wethLimitAmount,
+                uint8(FLASH_SOURCE.LEVERAGE_REBALANCE_DECREASE_DEBT),
+                abi.encodePacked(_usdcAmount)
+            );
+        }
+
+        _isValidLeverageRebalance();
+
+        emit LeverageRebalance(_isSellingUsdc, _usdcAmount, _wethLimitAmount);
+    }
+
+    /**
+     * @notice get current delta and bull CR ration in Euler
+     * @return delta and CR ratio
+     */
+    function getCurrentDeltaAndCollatRatio() external view returns (uint256, uint256) {
+        return _getCurrentDeltaAndCollatRatio();
+    }
+
+    /**
+     * @dev Uniswap V3 internal callback
+     * @param _uniFlashSwapData UniFlashswapCallbackData struct
+     */
+    function _uniFlashSwap(UniFlashswapCallbackData memory _uniFlashSwapData) internal override {
+        if (
+            FLASH_SOURCE(_uniFlashSwapData.callSource)
+                == FLASH_SOURCE.LEVERAGE_REBALANCE_INCREASE_DEBT
+        ) {
+            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(
+                IERC20(weth).balanceOf(address(this)), _uniFlashSwapData.amountToPay
+            );
+
+            IERC20(usdc).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
+        } else if (
+            FLASH_SOURCE(_uniFlashSwapData.callSource)
+                == FLASH_SOURCE.LEVERAGE_REBALANCE_DECREASE_DEBT
+        ) {
+            uint256 usdcToRepay = abi.decode(_uniFlashSwapData.callData, (uint256));
+            // Repay some USDC debt
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(
+                usdcToRepay, _uniFlashSwapData.amountToPay
+            );
+
+            IERC20(weth).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
+        } else if (
+            FLASH_SOURCE(_uniFlashSwapData.callSource)
+                == FLASH_SOURCE.GENERAL_SWAP
+        ) {
+            IERC20(_uniFlashSwapData.tokenIn).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
+        } else if (
+            FLASH_SOURCE(_uniFlashSwapData.callSource)
+                == FLASH_SOURCE.FULL_REBALANCE_BORROW_USDC_BUY_WETH
+        ) {
+            (uint256 wethToDeposit, uint256 usdcToBorrow) = abi.decode(_uniFlashSwapData.callData, (uint256, uint256));
+
+            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(wethToDeposit, usdcToBorrow);
```

Comment:
usdcToBorrow should just be _uniFlashSwapData.amountToPay (otherwise we may have extra borrow or not enough borrow!)